README.md
# Format As You Type

#### The simplest field formatting and masking tool for React.

Format fields the way you want using the input components you already use. Real-time. No hassle.

## Overview

* Install by running `npm i format-as-you-type` or `yarn add format-as-you-type`.
* Import to a React component using `import useFieldFormatter from 'format-as-you-type'`.
* Define a hook instance by passing in:
  1. A formatter (see below),
  2. A callback function that executes `onChange` (E.g., `const birthdayFormatter = useFieldFormatter(formatDate, setBirthday);`), and
  3. (Optionally) a custom ref prop to use on your field instead of the default `ref={}`.†
* The hook instance contains props to pass to your input component. Implement them by simply adding `{ ...yourConstantName }` at the end of the component's props.
<!-- * [See for yourself!]() -->

<p align="center">
  <img src="demo/demo.gif" width="400" alt="demo-gif" />
</p>
 
Want to see more examples? Keep reading!

† As you will see further below, there are some custom input components that won't behave as expected when using `ref={}`.

## Generic Formatters

While you can make your own formatters, there are a few that are ready to use:

* formatDate
* formatPhone
* formatCreditCard
* formatBigNumber
* formatPostalCodeCanada

More generic formatters will be added with upcoming versions. However, if you have an idea for a new one, feel free to suggest it by creating an issue.

See the __API__ header further below to learn more about each function's options and parameters.

## Make Your Own Formatter

Generic formatters not getting the job done? Make your own! Just import `formatString`. Here's an example:

```
import { formatString } from 'format-as-you-type';

const formatCustom = (newInput, options, /* any additional parameters */) => {
  // Any extra logic goes here...

  return formatString(newInput, '0000 \ 0000', options);
};
```

That same example can be implemented into the `useFieldFormatter` hook in a component like so:

```
const customFieldFormatter = useFieldFormatter(
  (newValue, options) => formatString(newInput, '0000 \ 0000', { ...options }),
  setCustomField,
);
```

Custom formats can either contain placeholder characters such as `A` for letters and `0` for numbers (e.g. `(000) 000-0000`), or actual letters and numbers (e.g. `(604) 123-4567`). All that matters is where separators (non-numbers and non-letters) are placed.

\***** The only rule is that __the first parameter must be the string being formatted__ and __the second parameter must accept an options argument__. \*****

## Implementing a Formatter in a Plain HTML Field

Implementing `useFieldFormatter` in a plan HTML field is easy:

```
import React from 'react';
import useFieldFormatter, { formatDate } from 'format-as-you-type';

const YourFormComponent = (props) => {
  const [birthday, setBirthday] = React.useState('');

  const birthdayFormatter = useFieldFormatter(formatDate, setBirthday);

  return (
    <form>
      /* ... */

      <label htmlFor="birthday">Birthday</label>
      <input
        id="birthday"
        value={birthday}
        { ...birthdayFormatter }
      />

      /* ... */
    </form>
  );
};
```

## Implementing a Formatter in a Custom Component

When using `useFieldFormatter` on a custom text field component, you may encounter warnings, errors, or other unpredictable behaviour. That's because, in the context of our example, spreading `birthdayFormatter` in a component yields the following by default:

```
<InputComponent
  /* other props */
  onChange={(event) => ourChangeHandler(event)}
  ref={ourInternalInputRef}
/>
```

Some custom input components don't work without using a different prop instead, such as `innerRef`, `inputRef`, etc., or wrapping the component in a `React.forwardRef`.

Since we can't (yet) predict what type of component the hook will be spread into, the simplest solution, at the moment, is to provide the ref prop as the third argument when initializing `useFieldFormatter`.

In Material-UI's case, this is `inputRef`. If you use another library and run into problems, give their documentation a read. For our example, though, we'll stick to Material-UI:

```
import React from 'react';
import useFieldFormatter, { formatDate } from 'format-as-you-type';
import TextField from '@mui/material/TextField';

const YourFormComponent = (props) => {
  const [birthday, setBirthday] = React.useState('');

  const birthdayFormatter = useFieldFormatter(formatDate, setBirthday, 'inputRef');

  return (
    <form>
      /* ... */

      <TextField
        label="Birthday"
        value={birthday}
        { ...birthdayFormatter }
      />

      /* ... */
    </form>
  );
};
```

## API

### useFieldFormatter

`useFieldFormatter(formatter, onChange, refProp = 'ref')`

##### Parameters

| Parameter | Description |
| --------- | ----------- |
| formatter | A generic or custom formatter function (see below) that takes in (1) a string and (2) an options object, in that order. |
| onChange | A callback function to update your field's value after if has been formatted. (This should have the structure `(newValue) => setValueSomewhere(newValue)`) |
| refProp | The prop to use to pass our internal prop to your input component. Is `'ref'` by default for plain HTML fields, but can be replaced with `inputRef`, `innerRef`, or whichever prop is instructed by your specific component/library. |

### formatDate

`formatDate(newInput, options, format = 'YYYY-MM-DD')`

### formatPhone

`formatPhone(newInput, options, format = '(000) 000-0000')`

### formatCreditCard

`formatCreditCard(newInput, options, infoType = 'NUMBER')`

##### Special parameters

| Parameter | Description | Options |
| --------- | ----------- | ------- |
| infoType | Whether to parse a credit card's number or its expiry date | NUMBER, EXPIRY_2 (`MM / YY`), EXPIRY_4 (`MM / YYYY`)|

##### Info

* This formatter is intended for credit cards with 16-digit numbers.
* If the first digit of a credit card's number is `3`, it will be formatted like an AMEX card. Otherwise, it'll be formatted in groups of four digits.
* If your form might take credit cards with more than 16 numbers, feel free to make a custom formatter using `formatString` (see further below).

### formatBigNumber

`formatBigNumber(newInput, options, locale)`

##### Special parameters

| Parameter | Description | Options |
| --------- | ----------- | ------- |
| locale | The region that determines how a number is formatted (e.g. 0,000,000 in English vs. 0.000.000 in French) | [en-US, fr-CA, etc...](https://stackoverflow.com/questions/3191664/list-of-all-locales-and-their-short-codes)† |

† This list will be converted to `type LocaleTypes = ...` once this package is converted to TypeScript very soon.

### formatPostalCodeCanada

`formatBigNumber(newInput, options)`

##### Info

This function simply formats strings to match the `A0A 0A0` format.

### formatString

`formatString(newInput, format, options)`

##### Special parameters

| Parameter | Description | Default Value |
| --------- | ----------- | ------- |
| options | Customization for the formatter. See documentation below. | `{}` |

##### Options

| Option | Description | Type (Default Value) |
| ------- | ----------- | ------------ |
| lettersAsSeparators | Treat letters as a separator. I.e. don't format them. | boolean (`false`) |
| numbersOnly | Doesn't allow letters in input. | boolean (`false`) |
| lettersOnly | Doesn't allow numbers in input. | boolean (`false`) |
| allCaps | Forces letters to be capitalized. | boolean (`false`) |
| customFormatter | Skip `formatString()` and use a custom formatter, such as JavaScript's built-in `number.toLocaleString()`, which is used in `formatBigNumber()`. | function that returns a string |


CHANGELOG.md
# Change Log

## [1.0.8] - 2022-03-03

* Browser auto-fill no longer causes an error.

## [1.0.7] - 2022-02-28

* Re-launch

demo/src/App.js
import React, { useState } from "react";
import { styled, createTheme, ThemeProvider } from "@mui/material/styles";
import CssBaseline from "@mui/material/CssBaseline";
import Container from "@mui/material/Container";
import Divider from "@mui/material/Divider";
import Dropdown from "./components/Dropdown";
import Footer from "./components/Footer";
import FormCard from "./components/FormCard";
import FormLabel from "@mui/material/FormLabel";
import FormRow from "./components/FormRow";
import Grid from "@mui/material/Grid";
import TextField from "./components/TextField";
import Typography from "@mui/material/Typography";
import { fieldSize } from "./constants";
import useFieldFormatter, {
  formatString,
  formatDate,
  formatPhone,
  formatCreditCard,
  formatPostalCodeCanada,
  formatBigNumber,
} from "format-as-you-type";

const App = () => {
  const [muiFieldValue, setMuiFieldValue] = useState('');
  const [htmlFieldValue, setHtmlFieldValue] = useState('');

  const [formatter, setFormatter] = useState('date');
  const [otherFormat, setOtherFormat] = useState('');

  const [creditCardMode, setCreditCardMode] = useState('NUMBER');

  const creditCardFormatter = (newValue, options) => {
    return formatCreditCard(newValue, options, creditCardMode);
  };

  const customFormatter = (newInput, options, format = otherFormat) => {
    return formatString(newInput, format, options);
  };

  const getFormatterFunction = (formatType = formatter) => {
    let formatterFunction;
    switch (formatType) {
      case 'other':
        formatterFunction = customFormatter;
        break;

      case 'creditCard':
        formatterFunction = creditCardFormatter;
        break;

      default:
        formatterFunction = formatters?.[formatType];
    }

    return formatterFunction;
  };

  const formatMuiField = useFieldFormatter(
    getFormatterFunction(),
    setMuiFieldValue,
    'inputRef'
  );

  const formatHtmlField = useFieldFormatter(
    getFormatterFunction(),
    setHtmlFieldValue
  );

  const handleFormatChange = (event) => {
    const newFormatterType = event.target.value;
    setFormatter(newFormatterType);

    const newFormatter = getFormatterFunction(newFormatterType);

    setMuiFieldValue( newFormatter(muiFieldValue) );
    setHtmlFieldValue( newFormatter(htmlFieldValue) );

    // Reset credit card mode
    creditCardMode !== 'NUMBER' && setCreditCardMode('NUMBER');
  };

  const handleOtherFormatChange = (event) => {
    const newFormat = event.target.value;
    setOtherFormat(newFormat);

    setMuiFieldValue( customFormatter(muiFieldValue, undefined, newFormat) );
    setHtmlFieldValue( customFormatter(htmlFieldValue, undefined, newFormat) );
  };

  const handleCreditCardModeChange = (event) => {
    setCreditCardMode(event.target.value);
    setMuiFieldValue('');
    setHtmlFieldValue('');
  };

  return (
    <>
      <CssBaseline />
      <ThemeProvider theme={CustomTheme}>
        {/* Header */}
        <Container
          maxWidth="sm"
          component="main"
          sx={{ padding: (theme) => theme.spacing(8, 0, 1) }}
        >
          <Typography component="h1" variant="h3" align="center" color="textPrimary" gutterBottom>
            Format As You Type.
          </Typography>
        </Container>
        {/* Main Content */}
        <Container maxWidth="md" component="main">
          <Grid
            container
            spacing={2}
            justifyContent="center"
            sx={{ mt: 1 }}
          >
            <FormCard>
              <FormRow>
                <FormLabel>
                  1. Pick a format.
                </FormLabel>
              </FormRow>
              <FormRow>
                <Dropdown
                  id="format"
                  value={formatter}
                  label="Format"
                  onChange={handleFormatChange}
                  choices={formatterDropdownChoices}
                />
              </FormRow>
              {/* Credit card mode field */}
              {formatter === 'creditCard' && (
                <FormRow>
                  <Dropdown
                    id="credit-card-mode"
                    value={creditCardMode}
                    label="Credit Card Info Type"
                    onChange={handleCreditCardModeChange}
                    choices={creditCardFormatModes}
                  />
                </FormRow>
              )}
              {/* Other format field */}
              {formatter === 'other' && (
                <FormRow>
                  <TextField
                    value={otherFormat}
                    label="Custom Format"
                    onChange={handleOtherFormatChange}
                    onClear={() => setOtherFormat('')}
                  />
                </FormRow>
              )}

              <FormRow>
                <Divider sx={{ mt: 1 }} />
              </FormRow>

              {/* MUI input field */}
              <FormRow>
                <FormLabel>
                  2. Test it out.
                </FormLabel>
                <TextField
                  value={muiFieldValue}
                  label="Custom Input Component"
                  onClear={() => setMuiFieldValue('')}
                  { ...formatMuiField }
                />
              </FormRow>
              {/* HTMLInputElement */}
              <FormRow>
                <Grid container alignItems="center">
                  <HTMLInputLabel
                    htmlFor="html-input-field"
                  >
                    Native HTML Input:&nbsp;
                  </HTMLInputLabel>
                  <FormRow>
                    <StyledHTMLInputField
                      id="html-input-field"
                      value={htmlFieldValue}
                      sx={{ mt: 1 }}
                      { ...formatHtmlField }
                    />
                  </FormRow>
                </Grid>
              </FormRow>
            </FormCard>
          </Grid>
        </Container>

        {/* Footer */}
        <Footer />
      </ThemeProvider>
    </>
  );
};

export default App;

// ********************************************************************
// SUPPLEMENTARY
const formatters = {
  date: formatDate,
  phone: formatPhone,
  postalCode: formatPostalCodeCanada,
  bigNumber: formatBigNumber,
};

const formatterDropdownChoices = {
  date: 'Date | YYYY-MM-DD',
  phone: 'Phone | (000) 000-0000',
  creditCard: 'Credit Card | 0000 0000 0000 0000',
  postalCode: 'Postal Code | A0A 0A0',
  bigNumber: 'Big Number | 0,000,000',
  other: 'Make Your Own Format',
};

const creditCardFormatModes = {
  NUMBER: 'Credit Card Number',
  EXPIRY_2: 'Expiry (MM / YY)',
  EXPIRY_4: 'Expiry (MM / YYYY)',
};

const CustomTheme = createTheme({
  typography: {
    fontFamily: ['-apple-system', 'BlinkMacSystemFont', '"Segoe UI"', '"Helvetica Neue"',
      'Arial', 'sans-serif', '"Apple Color Emoji"', '"Segoe UI Emoji"', '"Segoe UI Symbol"',
    ].join(','),
  }
});

const StyledHTMLInputField = styled('input')({
  marginRight: 2,
  fontSize: '1rem',
  fontWeight: 400,
  lineHeight: fieldSize === 'small' ? '2.2rem' : '2.6rem',
  borderRadius: 5,
  borderStyle: 'solid',
  borderWidth: '1.25px',
  paddingLeft: '0.9rem',
  paddingRight: '0.9rem',
  width: '100%',
});

const HTMLInputLabel = styled('label')({
  color: 'dimgrey',
  fontSize: '0.95rem',
});
